src/Operation/Rounding.php

```diff
@@ -0,0 +1,420 @@
+         * if D = 5 and E is odd and X is negative, roundHalfUp(X, P) = floor(X, P)
+         */
+
+        $fractionalPart = $number->getFractionalPart();
+
+        $digit = (int) $fractionalPart[$precision];
+
+        if ($digit !== 5) {
+            return $this->roundHalfUp($number, $precision);
+        }
+
+        // retrieve the digit to the left of it
+        if ($precision === 0) {
+            $referenceDigit = (int) substr($number->getIntegerPart(), -1);
+        } else {
+            $referenceDigit = (int) $fractionalPart[$precision - 1];
+        }
+
+        // truncate if even
+        $isEven = $referenceDigit % 2 === 0;
+        if ($isEven) {
+            return $this->truncate($number, $precision);
+        }
+
+        // round away from zero
+        $method = ($number->isPositive()) ? self::ROUND_CEIL : self::ROUND_FLOOR;
+        return $this->compute($number, $precision, $method);
+    }
+
+    /**
+     * Rounds the number according to the digit D located at precision P.
+     * - It rounds away from zero if D >= $halfwayValue
+     * - It rounds towards zero if D < $halfWayValue
+     *
+     * @param DecimalNumber $number Number to round
+     * @param int $precision Maximum number of decimals
+     * @param int $halfwayValue Threshold upon which the rounding will be performed
+     *  away from zero instead of towards zero.
+     *
+     * @return DecimalNumber
+     */
+    private function roundHalf(DecimalNumber $number, $precision, $halfwayValue)
+    {
+        $precision = $this->sanitizePrecision($precision);
+
+        if ($number->getPrecision() <= $precision) {
+            return $number;
+        }
+
+        /**
+         * The principle for roundHalf is the following:
+         *
+         * let X = number to round
+         *     P = number of decimal digits that we want
+         *     D = digit from the fractional part at index P
+         *     Y = digit considered as the half-way value on which we round up (usually 5 or 6)
+         *
+         * if D >= Y, roundHalf(X, P) = ceil(X, P)
+         * if D < Y, roundHalf(X, P) = truncate(X, P)
+         */
+
+        $fractionalPart = $number->getFractionalPart();
+
+        $digit = (int) $fractionalPart[$precision];
+        if ($digit >= $halfwayValue) {
+            // round away from zero
+            $mode = ($number->isPositive()) ? self::ROUND_CEIL : self::ROUND_FLOOR;
+            return $this->compute($number, $precision, $mode);
+        }
+
+        // round towards zero
+        return $this->truncate($number, $precision);
+    }
+
+    /**
+     * Ensures that precision is a positive int
+     *
+     * @param mixed $precision
+     *
+     * @return int Precision
+     *
+     * @throws \InvalidArgumentException if precision is not a positive integer
+     */
+    private function sanitizePrecision($precision)
+    {
+        if (!is_numeric($precision) || $precision < 0) {
```

Comment:
> Like inverting negative values ?

Not necessarily (it depends on the sanitization constraints)

> As is, this is a mix between validation and sanitization.

Pretty much, yeah 🙂 

At the start, the idea was simply to make sure that precision was a valid int value (`>= 0`). At that point, I had a validation.

Then I made the design choice of making `Rounding` methods flexible enough to let them receive `'1'` instead of `1`, while still considering `'-1'` and `'true'` as invalid values. This prevented me from coercing the value to int directly because it would mean accepting any int-coercible value as a valid.

_(As a side note, the `is_numeric()` check has the side effect of accepting values like `1.12345`, `0x1`', `1e0` and `'1e0'`, but it's not a big deal)_

So, I had a validation and type coercion for certain kinds of values. I needed a single method that would allow me to consume the parameter as an int without having to worry about whether it is valid or not.

"Sanitize" was the best word I could find, but I'm always open if you find a better one 🙂